### PR TITLE
[PLAT-46720] Support mlflow run export

### PR DESCRIPTION
**Pull request by kevinkim9264**
This PR supports the following command:
```
python3 export_db.py --profile dogfood --mlflow-runs --use-checkpoint --num-parallel 8

022-03-08,17:07:00;INFO;Complete MLflow Runs Export Time: 0:07:13.251757
```

Given the scale of the runs (total runs and runs per experimentId. total 220k runs for shard-dogfood), we store the run object data into sqlite. 

Unit test: `python3 -m unittest dbclient/test/MLFlowClientTest.py`
Manual test: 
```
python3 export_db.py --profile dogfood --mlflow-runs --use-checkpoint --num-parallel 1  (take 20 minutes)
python3 export_db.py --profile dogfood --mlflow-runs --use-checkpoint --num-parallel 8 (take 7 minutes)
python3 export_db.py --profile dogfood --mlflow-runs --use-checkpoint --num-parallel 10 (take 5 minutes)
```

Check the total size of the data to make sure different number of threads result in same number of data:
```
con_1 = sqlite3.connect("logs/mlflow_runs.db_1_thread", timeout=10)
con_1 = sqlite3.connect("logs/mlflow_runs.db_1_thread", timeout=10)
con_1 = sqlite3.connect("logs/mlflow_runs.db_1_thread", timeout=10)
cur_1 = con_1.execute('''SELECT COUNT(*) FROM runs''')
cur_1.fetchone()
(217947,)

con_2 = sqlite3.connect("logs/mlflow_runs.db_10_thread", timeout=10)
con_2 = sqlite3.connect("logs/mlflow_runs.db_10_thread", timeout=10)
cur_2 = con_2.execute('''SELECT COUNT(*) FROM runs''')
cur_2.fetchone()
(217947,)

con_3 = sqlite3.connect("logs/mlflow_runs.db_8_thread", timeout=10)
cur_3 = con_3.execute('''SELECT COUNT(*) FROM runs''')
cur_3.fetchone()
(217947,)
```

### CODE REVIEW

File: dbclient/MLFlowClient.py [hunk 1]
```diff
@@ -11,13 +11,118 @@
...
         self._checkpoint_service = checkpoint_service
         self.export_dir = configs['export_dir']
         self.client = MlflowClient(f"databricks://{configs['profile']}")
 
+    def export_mlflow_runs(self, log_sql_file='mlflow_runs.db', experiment_log='mlflow_experiments.log', num_parallel=4):
+        """
+        Exports the Mlflow run objects. This can be run only after export_mlflow_experiments is complete.
+        Unlike other objects, we save the data into sqlite tables, given the possible scale of runs objects.
+        """
+        print(f"num_parallel: {num_parallel}")
```
Comment (kevinkim9264) [inline dbclient/MLFlowClient.py:27]: TODO(kevin): remove

**Review by gregwood-db — APPROVED** [review]

File: dbclient/MLFlowClient.py [hunk 2]
```diff
@@ -35,7 +140,7 @@ def export_mlflow_experiments(self, log_file='mlflow_experiments.log', log_dir=N
         logging.info("Complete MLflow Experiments Export Time: " + str(timedelta(seconds=end - start)))
 
     def import_mlflow_experiments(self, log_file='mlflow_experiments.log', id_map_file='mlflow_experiments_id_map.log',
-                                  log_dir=None, num_parallel=4):
+                                  log_dir=None, num_parallel=6):
```
Comment (smurching) [inline dbclient/MLFlowClient.py:143]: To confirm, this change is unrelated to run import right? Is it intended?

File: dbclient/test/MLFlowClientTest.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        runs_dict[run_id] = run
+        return run
+
+    def _insert_run_data(self, run):
+        con = sqlite3.connect(MLFLOW_TEST_FILE, timeout=30)
+        with con:
+            MLFlowClient._save_run_data_to_sql(con, run)
+
+
+    def test_save_run_data_to_sql(self):
```
Comment (smurching) [inline dbclient/test/MLFlowClientTest.py:49]: Could we also add a test that proves the idempotency of `MlflowClient._save_run_data_to_sql`, i.e. in case we fail to export an experiment's runs partway and so retry exporting its runs + re-export some runs?

Comment (smurching) [reply dbclient/test/MLFlowClientTest.py:49]: It looks like it's already idempotent due to the `REPLACE INTO` syntax, but just thinking a test would be nice to have here

**Review by smurching — COMMENTED** [review]
Mostly LGTM, just a few comments. Those performance numbers are pretty encouraging!

File: dbclient/MLFlowClient.py (hunk 2 above)
Comment (kevinkim9264) [reply dbclient/MLFlowClient.py:143]: √ reverting

File: dbclient/test/MLFlowClientTest.py (hunk 3 above)
Comment (kevinkim9264) [reply dbclient/test/MLFlowClientTest.py:49]: good point. added the test!

File: dbclient/MLFlowClient.py [hunk 4]
```diff
@@ -11,13 +11,117 @@
...
+        experiments_logfile = self.export_dir + experiment_log
+        mlflow_runs_checkpointer = self._checkpoint_service.get_checkpoint_key_set(
+            wmconstants.WM_EXPORT, wmconstants.MLFLOW_RUNS
+        )
+
+        start = timer()
+        con = sqlite3.connect(self.export_dir + log_sql_file)
+        with con:
+            con.execute('''
+              CREATE TABLE IF NOT EXISTS runs (id TEXT UNIQUE, start_time INT, run_obj TEXT)
```
Comment (souravkhandelwal09) [inline dbclient/MLFlowClient.py:36]: is the run id unique across experiments ?

File: dbclient/MLFlowClient.py [hunk 5]
```diff
@@ -11,13 +11,117 @@
...
+                is_there_exception = True
+            else:
+                # With multi-threading, we need to give enough timeout for each thread's connection to be made
+                # Upon DatabaseLock timeout exception, one can simply rerun the command since the checkpoint saves the
+                # progress
+                con = sqlite3.connect(self.export_dir + log_sql_file, timeout=180)
+                with con:
+                    for run in runs:
+                        self._save_run_data_to_sql(con, run)
+                con.close()
```
Comment (souravkhandelwal09) [inline dbclient/MLFlowClient.py:104]: there could be errors in this section of code as well, we need to log them to error_logger

File: dbclient/MLFlowClient.py [hunk 6]
```diff
@@ -11,13 +11,117 @@
...
+    @classmethod
+    def _save_run_data_to_sql(cls, con, run):
+        run_id = run.info.run_id
+        start_time = run.info.start_time
+        run_object = {
+            "info": dict(run.info),
+            "metrics": dict(run.data.metrics),
+            "params": dict(run.data.params),
+            "tags": dict(run.data.tags)
+        }
```
Comment (souravkhandelwal09) [inline dbclient/MLFlowClient.py:120]: is this format needed for import ?

**Review by souravkhandelwal09 — COMMENTED** [review]
mostly looks good, minor comments

File: dbclient/MLFlowClient.py (hunk 4 above)
Comment (kevinkim9264) [reply dbclient/MLFlowClient.py:36]: correct

File: dbclient/MLFlowClient.py (hunk 6 above)
Comment (kevinkim9264) [reply dbclient/MLFlowClient.py:120]: Not this particular format, but we do need info, metrics, params, tags separate objects.

File: dbclient/MLFlowClient.py (hunk 5 above)
Comment (kevinkim9264) [reply dbclient/MLFlowClient.py:104]: I think we aim to only error_logger the API errors. For the other run time errors, etc, we would let the script just directly raise and exit (instead of waiting for the whole task to finish, which is done by error_logger)

**Review by smurching — APPROVED** [review]
LGTM

File: dbclient/MLFlowClient.py (hunk 5 above)
Comment (souravkhandelwal09) [reply dbclient/MLFlowClient.py:104]: ok sounds good

**Review by souravkhandelwal09 — APPROVED** [review]
